File: api/gateway/native/src/routes/v1/bookshelf.ts
import express, { NextFunction, Request, Response } from 'express'
import {
  deleteBookshelf,
  getBookshelf,
  getUser,
  listBookshelf,
  listUserWithUserIds,
  readBookshelf,
  readingBookshelf,
  releaseBookshelf,
  stackBookshelf,
  wantBookshelf,
} from '~/api'
import { BookStatus } from '~/types/book'
import { GrpcError } from '~/types/exception'
import {
  IDeleteBookshelfInput,
  IGetBookshelfInput,
  IGetUserInput,
  IListBookshelfInput,
  IListUserByUserIdsInput,
  IReadBookshelfInput,
  IReadingBookshelfInput,
  IReleaseBookshelfInput,
  IStackBookshelfInput,
  IWantBookshelfInput,
} from '~/types/input'
import {
  IBookshelfListOutput,
  IBookshelfListOutputAuthor,
  IBookshelfListOutputBookshelf,
  IBookshelfOutput,
  IBookshelfOutputAuthor,
  IBookshelfOutputReview,
  IUserHashOutput,
} from '~/types/output'
import { IReadBookshelfRequest } from '~/types/request'
import {
  IBookshelfListResponse,
  IBookshelfListResponseBook,
  IBookshelfListResponseBookshelf,
  IBookshelfResponse,
  IBookshelfResponseBookshelf,
  IBookshelfResponseReview,
  IBookshelfResponseUserOnReview,
} from '~/types/response'

const router = express.Router()

router.get(
  '/v1/users/:userId/books',
  async (req: Request, res: Response<IBookshelfListResponse>, next: NextFunction): Promise<void> => {
    const { userId } = req.params
    const { limit, offset } = req.query as { [key: string]: string }

    const userInput: IGetUserInput = {
      id: userId,
    }

    await getUser(req, userInput)
      .then(async () => {
        const bookshelfInput: IListBookshelfInput = {
          userId,
          limit: Number(limit) || 100,
          offset: Number(offset) || 0,
        }

        return listBookshelf(req, bookshelfInput)
      })
      .then((output: IBookshelfListOutput) => {
        const response: IBookshelfListResponse = setBookshelfListResponse(output)
        res.status(200).json(response)
      })
      .catch((err: GrpcError) => next(err))
  }
)

router.get(
  '/v1/users/:userId/books/:bookId',
  async (req: Request, res: Response<IBookshelfResponse>, next: NextFunction): Promise<void> => {
    const { userId, bookId } = req.params

    const userInput: IGetUserInput = {
      id: userId,
    }

    await getUser(req, userInput)
      .then(async () => {
        const bookshelfInput: IGetBookshelfInput = {
          userId,
          bookId: Number(bookId) || 0,
        }

        return getBookshelf(req, bookshelfInput)
      })
      .then(async (bookshelfOutput: IBookshelfOutput) => {
        let userIds: string[] = []
        bookshelfOutput.book.reviews.forEach((rv: IBookshelfOutputReview) => {
          userIds = userIds.concat(rv.userId)
        })

        const userListInput: IListUserByUserIdsInput = {
          ids: userIds,
        }

        return listUserWithUserIds(req, userListInput)
          .then((usersOutput: IUserHashOutput) => {
            return setBookshelfResponse(bookshelfOutput, usersOutput)
          })
          .catch(() => {
            return setBookshelfResponse(bookshelfOutput, {})
          })
      })
      .then((response: IBookshelfResponse) => {
        res.status(200).json(response)
      })
      .catch((err: GrpcError) => next(err))
  }
)

router.post(
  '/v1/users/:userId/books/:bookId/read',
  async (req: Request, res: Response<IBookshelfResponse>, next: NextFunction): Promise<void> => {
    const { userId, bookId } = req.params
    const { readOn, impression } = req.body as IReadBookshelfRequest

    const userInput: IGetUserInput = {
      id: userId,
    }

    await getUser(req, userInput)
      .then(async () => {
        const bookshelfInput: IReadBookshelfInput = {
          userId,
          bookId: Number(bookId) || 0,
          impression,
          readOn,
        }

        return readBookshelf(req, bookshelfInput)
      })
      .then((output: IBookshelfOutput) => {
        const response: IBookshelfResponse = setBookshelfResponse(output, {})
        res.status(200).json(response)
      })
      .catch((err: GrpcError) => next(err))
  }
)

router.post(
  '/v1/users/:userId/books/:bookId/reading',
  async (req: Request, res: Response<IBookshelfResponse>, next: NextFunction): Promise<void> => {
    const { userId, bookId } = req.params

    const userInput: IGetUserInput = {
      id: userId,
    }

    await getUser(req, userInput)
      .then(async () => {
        const bookshelfInput: IReadingBookshelfInput = {
          userId,
          bookId: Number(bookId) || 0,
        }

        return readingBookshelf(req, bookshelfInput)
      })
      .then((output: IBookshelfOutput) => {
        const response: IBookshelfResponse = setBookshelfResponse(output, {})
        res.status(200).json(response)
      })
      .catch((err: GrpcError) => next(err))
  }
)

router.post(
  '/v1/users/:userId/books/:bookId/stack',
  async (req: Request, res: Response<IBookshelfResponse>, next: NextFunction): Promise<void> => {
    const { userId, bookId } = req.params

    const userInput: IGetUserInput = {
      id: userId,
    }

    await getUser(req, userInput)
      .then(async () => {
        const bookshelfInput: IStackBookshelfInput = {
          userId,
          bookId: Number(bookId) || 0,
        }

        return stackBookshelf(req, bookshelfInput)
      })
      .then((output: IBookshelfOutput) => {
        const response: IBookshelfResponse = setBookshelfResponse(output, {})
        res.status(200).json(response)
      })
      .catch((err: GrpcError) => next(err))
  }
)

router.post(
  '/v1/users/:userId/books/:bookId/want',
  async (req: Request, res: Response<IBookshelfResponse>, next: NextFunction): Promise<void> => {
    const { userId, bookId } = req.params

    const userInput: IGetUserInput = {
      id: userId,
    }

    await getUser(req, userInput)
      .then(async () => {
        const bookshelfInput: IWantBookshelfInput = {
          userId,
          bookId: Number(bookId) || 0,
        }

        return wantBookshelf(req, bookshelfInput)
      })
      .then((output: IBookshelfOutput) => {
        const response: IBookshelfResponse = setBookshelfResponse(output, {})
        res.status(200).json(response)
      })
      .catch((err: GrpcError) => next(err))
  }
)

router.post(
  '/v1/users/:userId/books/:bookId/release',
  async (req: Request, res: Response<IBookshelfResponse>, next: NextFunction): Promise<void> => {
    const { userId, bookId } = req.params

    const userInput: IGetUserInput = {
      id: userId,
    }

    await getUser(req, userInput)
      .then(async () => {
        const bookshelfInput: IReleaseBookshelfInput = {
          userId,
          bookId: Number(bookId) || 0,
        }

        return releaseBookshelf(req, bookshelfInput)
      })
      .then((output: IBookshelfOutput) => {
        const response: IBookshelfResponse = setBookshelfResponse(output, {})
        res.status(200).json(response)
      })
      .catch((err: GrpcError) => next(err))
  }
)

router.delete(
  '/v1/users/:userId/books/:bookId',
  async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    const { userId, bookId } = req.params

    const userInput: IGetUserInput = {
      id: userId,
    }

    await getUser(req, userInput)
      .then(async () => {
        const bookshelfInput: IDeleteBookshelfInput = {
          userId,
          bookId: Number(bookId) || 0,
        }

        return deleteBookshelf(req, bookshelfInput)
      })
      .then(() => {
        res.status(200).json({ status: 'ok' })
      })
      .catch((err: GrpcError) => next(err))
  }
)

function setBookshelfResponse(bookshelfOutput: IBookshelfOutput, usersOutput: IUserHashOutput): IBookshelfResponse {
  const authorNames: string[] = bookshelfOutput.book.authors.map((item: IBookshelfOutputAuthor) => {
    return item.name
  })

  const authorNameKanas: string[] = bookshelfOutput.book.authors.map((item: IBookshelfOutputAuthor) => {
    return item.nameKana
  })

  const reviews = bookshelfOutput.book.reviews.map(
    (item: IBookshelfOutputReview): IBookshelfResponseReview => {
      const user: IBookshelfResponseUserOnReview = {
        id: item.userId,
        username: '',
        thumbnailUrl: '',
      }

      if (usersOutput[item.userId]) {
        user.username = usersOutput[item.userId].username
        user.thumbnailUrl = usersOutput[item.userId].thumbnailUrl
      }

      return {
        id: item.id,
        impression: item.impression,
        createdAt: item.createdAt,
        updatedAt: item.updatedAt,
        user,
      }
    }
  )

  const bookshelf: IBookshelfResponseBookshelf = {
    id: bookshelfOutput.id,
    status: BookStatus[bookshelfOutput.status],
    readOn: bookshelfOutput.readOn,
    impression: bookshelfOutput.myReview?.impression || '',
    createdAt: bookshelfOutput.createdAt,
    updatedAt: bookshelfOutput.updatedAt,
  }

  const response: IBookshelfResponse = {
    id: bookshelfOutput.book.id,
    title: bookshelfOutput.book.title,
    titleKana: bookshelfOutput.book.titleKana,
    description: bookshelfOutput.book.description,
    isbn: bookshelfOutput.book.isbn,
    publisher: bookshelfOutput.book.publisher,
    publishedOn: bookshelfOutput.book.publishedOn,
    thumbnailUrl: bookshelfOutput.book.thumbnailUrl,
    rakutenUrl: bookshelfOutput.book.rakutenUrl,
    size: bookshelfOutput.book.rakutenSize,
    author: authorNames.join('/'),
    authorKana: authorNameKanas.join('/'),
    createdAt: bookshelfOutput.book.createdAt,
    updatedAt: bookshelfOutput.book.updatedAt,
    bookshelf,
    reviews,
    reviewLimit: bookshelfOutput.book.reviewLimit,
    reviewOffset: bookshelfOutput.book.reviewOffset,
    reviewTotal: bookshelfOutput.book.reviewTotal,
  }

  return response
}

function setBookshelfListResponse(output: IBookshelfListOutput): IBookshelfListResponse {
  const books: IBookshelfListResponseBook[] = output.bookshelves.map((bs: IBookshelfListOutputBookshelf) => {
    const authorNames: string[] = bs.book.authors.map((item: IBookshelfListOutputAuthor) => {
      return item.name
    })

    const authorNameKanas: string[] = bs.book.authors.map((item: IBookshelfListOutputAuthor) => {
      return item.nameKana
    })

    const bookshelf: IBookshelfListResponseBookshelf = {
      id: bs.id,
      status: BookStatus[bs.status],
      readOn: bs.readOn,
      impression: bs.myReview?.impression || '',
      createdAt: bs.createdAt,
      updatedAt: bs.updatedAt,
    }

    const book: IBookshelfListResponseBook = {
      id: bs.book.id,
      title: bs.book.title,
      titleKana: bs.book.titleKana,
      description: bs.book.description,
      isbn: bs.book.isbn,
      publisher: bs.book.publisher,
      publishedOn: bs.book.publishedOn,
      thumbnailUrl: bs.book.thumbnailUrl,
      rakutenUrl: bs.book.rakutenUrl,
      size: bs.book.rakutenSize,
      author: authorNames.join('/'),
      authorKana: authorNameKanas.join('/'),
      createdAt: bs.book.createdAt,
      updatedAt: bs.book.updatedAt,
      bookshelf,
    }

    return book
  })

  const response: IBookshelfListResponse = {
    books,
    limit: output.limit,
    offset: output.offset,
    total: output.total,
  }

  return response
}

export default router

File: api/gateway/native/internal/response/v1/bookshelf.go
package v1

import (
	"strings"

	"github.com/calmato/gran-book/api/gateway/native/internal/entity"
)

// 本棚の書籍情報
type BookshelfResponse struct {
	ID           int64               `json:"id"`           // 書籍ID
	Title        string              `json:"title"`        // タイトル
	TitleKana    string              `json:"titleKana"`    // タイトル(かな)
	Description  string              `json:"description"`  // 説明
	Isbn         string              `json:"isbn"`         // ISBN
	Publisher    string              `json:"publisher"`    // 出版社名
	PublishedOn  string              `json:"publishedOn"`  // 出版日
	ThumbnailURL string              `json:"thumbnailUrl"` // サムネイルURL
	RakutenURL   string              `json:"rakutenUrl"`   // 楽天ショップURL
	Size         string              `json:"size"`         // 楽天書籍サイズ
	Author       string              `json:"author"`       // 著者名一覧
	AuthorKana   string              `json:"author_kana"`  // 著者名一覧(かな)
	Bookshelf    *bookshelfBookshelf `json:"bookshelf"`    // ユーザーの本棚情報
	CreatedAt    string              `json:"createdAt"`    // 登録日時
	UpdatedAt    string              `json:"updatedAt"`    // 更新日時
}

func NewBookshelfResponse(bs *entity.Bookshelf, b *entity.Book) *BookshelfResponse {
	return &BookshelfResponse{
		ID:           b.Id,
		Title:        b.Title,
		TitleKana:    b.TitleKana,
		Description:  b.Description,
		Isbn:         b.Isbn,
		Publisher:    b.Publisher,
		PublishedOn:  b.PublishedOn,
		ThumbnailURL: b.ThumbnailUrl,
		RakutenURL:   b.RakutenUrl,
		Size:         b.RakutenSize,
		Author:       strings.Join(b.AuthorNames(), "/"),
		AuthorKana:   strings.Join(b.AuthorNameKanas(), "/"),
		CreatedAt:    b.CreatedAt,
		UpdatedAt:    b.UpdatedAt,
		Bookshelf:    newBookshelfBookshelf(bs),
	}
}

type bookshelfBookshelf struct {
	ID         int64  `json:"id"`         // 本棚ID
	Status     string `json:"status"`     // 読書ステータス
	ReadOn     string `json:"readOn"`     // 読み終えた日
	Impression string `json:"impression"` // 感想
	CreatedAt  string `json:"createdAt"`  // 登録日時
	UpdatedAt  string `json:"updatedAt"`  // 更新日時
}

func newBookshelfBookshelf(bs *entity.Bookshelf) *bookshelfBookshelf {
	return &bookshelfBookshelf{
		ID:         bs.Id,
		Status:     bs.Status().Name(),
		ReadOn:     bs.ReadOn,
		Impression: "",
		CreatedAt:  bs.CreatedAt,
		UpdatedAt:  bs.UpdatedAt,
	}
}

// 本棚の書籍一覧
type BookshelfListResponse struct {
	Books  []*bookshelfListBook `json:"booksList"` // 書籍一覧
	Limit  int64                `json:"limit"`     // 取得上限数
	Offset int64                `json:"offset"`    // 取得開始位置
	Total  int64                `json:"total"`     // 検索一致数
}

func NewBookshelfListResponse(
	bss entity.Bookshelves, bm map[int64]*entity.Book, limit, offset, total int64,
) *BookshelfListResponse {
	return &BookshelfListResponse{
		Books:  newBookshelfListBooks(bss, bm),
		Limit:  limit,
		Offset: offset,
		Total:  total,
	}
}

type bookshelfListBook struct {
	ID           int64                   `json:"id"`           // 書籍ID
	Title        string                  `json:"title"`        // タイトル
	TitleKana    string                  `json:"titleKana"`    // タイトル(かな)
	Description  string                  `json:"description"`  // 説明
	Isbn         string                  `json:"isbn"`         // ISBN
	Publisher    string                  `json:"publisher"`    // 出版社名
	PublishedOn  string                  `json:"publishedOn"`  // 出版日
	ThumbnailURL string                  `json:"thumbnailUrl"` // サムネイルURL
	RakutenURL   string                  `json:"rakutenUrl"`   // 楽天ショップURL
	Size         string                  `json:"size"`         // 楽天書籍サイズ
	Author       string                  `json:"author"`       // 著者名一覧
	AuthorKana   string                  `json:"authorKana"`   // 著者名一覧(かな)
	Bookshelf    *bookshelfListBookshelf `json:"bookshelf"`    // ユーザーの本棚情報
	CreatedAt    string                  `json:"createdAt"`    // 登録日時
	UpdatedAt    string                  `json:"updatedAt"`    // 更新日時
}

func newBookshelfListBook(bs *entity.Bookshelf, b *entity.Book) *bookshelfListBook {
	return &bookshelfListBook{
		ID:           b.Id,
		Title:        b.Title,
		TitleKana:    b.TitleKana,
		Description:  b.Description,
		Isbn:         b.Isbn,
		Publisher:    b.Publisher,
		PublishedOn:  b.PublishedOn,
		ThumbnailURL: b.ThumbnailUrl,
		RakutenURL:   b.RakutenUrl,
		Size:         b.RakutenSize,
		Author:       strings.Join(b.AuthorNames(), "/"),
		AuthorKana:   strings.Join(b.AuthorNameKanas(), "/"),
		CreatedAt:    b.CreatedAt,
		UpdatedAt:    b.UpdatedAt,
		Bookshelf:    newBookshelfListBookshelf(bs),
	}
}

func newBookshelfListBooks(bss entity.Bookshelves, bm map[int64]*entity.Book) []*bookshelfListBook {
	res := make([]*bookshelfListBook, 0, len(bss))
	for i, bs := range bss {
		b, ok := bm[bs.BookId]
		if !ok {
			continue
		}

		res[i] = newBookshelfListBook(bs, b)
	}
	return res
}

type bookshelfListBookshelf struct {
	ID        int64  `json:"id"`        // 本棚ID
	Status    string `json:"status"`    // 読書ステータス
	ReadOn    string `json:"readOn"`    // 読み終えた日
	CreatedAt string `json:"createdAt"` // 登録日時
	UpdatedAt string `json:"updatedAt"` // 更新日時
}

func newBookshelfListBookshelf(bs *entity.Bookshelf) *bookshelfListBookshelf {
	return &bookshelfListBookshelf{
		ID:        bs.Id,
		Status:    bs.Status().Name(),
		ReadOn:    bs.ReadOn,
		CreatedAt: bs.CreatedAt,
		UpdatedAt: bs.UpdatedAt,
	}
}
